### ISSUE TITLE
Problem with a large db

### ISSUE THREAD

**Issue by ribelo** [juji-io/datalevin#10]
When the db file reaches about 100 Mb, error occurs during the transaction

```clj
Execution error (ExceptionInfo) at datalevin.lmdb.LMDB/transact (lmdb.clj:665).
Fail to transact to LMDB: "\"BufferOverflow: trying to put 8 bytes while 7 remaining in the ByteBuffer.\""
```

EDIT:

Limit is not 100Mb. After deleting db files and transact the data again, it works.

**Comment by huahaiy** [comment]
Thanks for reporting. I will look into this later.

**Comment by huahaiy** [comment]
Released 0.2.16 to have more detailed error message on buffer overflow, so we can better pinpoint where the error come from. Please try this version and file issue with the new errors. Closing this one.

**Comment by ribelo** [comment]
Unfortunately, but the problem is still there.

I tried to isolate him. I can't. I was loading 10gb of random data into the DB. Everything's working. 

But sometimes when I work on my data it spits out the error.

```Fail to transact to LMDB: "Error putting r/w key buffer of datalevin/eav." \\\"BufferOverflow: trying to put 1 bytes while 0 remaining in the ByteBuffer."```

I'd love to find out where it comes from, but I don't know how. What else can I do to help?

**Comment by huahaiy** [comment]
From the error, it appears that the error is thrown when writing the index key when writing a large value. I don't think the size of the database has anything to do with the error. It's the particular type of data is causing the error, so it is possible to isolate it by simply printing out each piece of data before transacting it. If you could post the code, I can further investigate.

**Comment by ribelo** [comment]
I tried. However, when I restarted REPL, the data that did not want to be transacted before, were transacted without any problems. I still work on the same data and everything works 99% of the time. I also don't have any weird data types in the DB. Just strings and numbers.

DB has this scheme:

```clojure
:dc.warehouse.product/market-id      {:db/valueType :db.type/string}
:dc.warehouse.product/ean            {:db/valueType :db.type/string}
:dc.warehouse.product/name           {:db/valueType :db.type/string}
:dc.warehouse.product/id             {:db/valueType :db.type/string}
:dc.warehouse.product/group-id       {:db/valueType :db.type/string}
:dc.warehouse.product/unit           {:db/valueType :db.type/string}
:dc.warehouse.product/rate-of-sales  {:db/valueType :db.type/double}
:dc.warehouse.product/purchase-price {:db/valueType :db.type/double}
:dc.warehouse.product/vat            {:db/valueType :db.type/double}
:dc.warehouse.product/gross-price    {:db/valueType :db.type/double}
:dc.warehouse.product/stock          {:db/valueType :db.type/double}
:dc.warehouse.product/margin         {:db/valueType :db.type/double}
```
In a big simplification, this is what the code looks like. `put-txs` is always a vector of maps with the given scheme.

```clojure
(defn update-market! [market-id file-path]
  (let [market-id (-> market-id (name) (str/lower))
        put-txs   (read-market-warehouse file-path)]
    (let [del-txs (->> (d/q '[:find [?e ...]
                              :in $ ?market-id
                              :where
                              [?e :dc.warehouse.product/ean]
                              [?e :dc.warehouse/market-id ?market-id]]
                         @db/conn market-id)
                       (mapv (fn [?e] [:db/retractEntity ?e])))]
      (d/transact! db/conn (into del-txs put-txs)))))
```

Surely the next time I meet this bug, I will spend more time on it.

**Comment by huahaiy** [comment]
Thanks for the information. It looks like the cause is uncleared buffers. I have released 0.2.18 to hopefully address the issue.  Please give it a try.

**Comment by huahaiy** [comment]
Please reopen if the problem persists.
